Feat/scene

mateuseap opened the pull request:
---
title: Feat/scene
labels: enhancement

---

## Related issues
* Closes #5

## What I did
> Add `Scene` header, source code and test files.

## Properties
- `std::map<int, Light>`: scene lights;
- `int lightsCurrentIndex`: current index of scene lights map;
- `Color environmentColor`: scene environment color;
- `std::map<int, Box> objects`: scene objects;
- `int objectsCurrentIndex`: current index of scene objects map.

## Methods
- `getLights`: returns `std::map<int, Light> lights` attribute;
- `getEnvironmentColor`: returns `Color environmentColor` attribute;
- `getObjects`: returns `std::map<int, Box> objects` attribute;
- `addLight`: add a new `Light` to `std::map<int, Light>` and returns its index;
- `removeLight`: removes a `Light` at the given index;
- `setEnvironmentColor`: sets `Color environmentColor` attribute;
- `addObject`: add a new `Box` to `std::map<int, Box>` and returns its index;
- `removeObject`: removes a `Box` at the given index.

## Additional notes
- I've add add **equal (==)** operator, **not equal (!=)** and `Light()` empty constructor in `Light` class;
- I've add **equal (==)** and **not equal (!=)** operators in `Color` class too, since it is used in `Light` class, I've created these operators to make the comparisons easier.

@kinhosz
@mateuseap

mateuseap committed:
feat(include/graphic/color.h): add equal and not equal operators

mateuseap committed:
feat(src/graphic/color.cpp): add equal and not equal operators source code

mateuseap committed:
feat(include/entity/light.h): add equal and not equal operators

mateuseap committed:
feat(src/entity/light.cpp): add equal and not equal operators source code

mateuseap committed:
feat(include/entity): add 'Scene' header file

mateuseap committed:
feat(src/entity): add 'Scene' source code file

mateuseap committed:
feat(tests/entity): add 'Scene' test file

mateuseap commented:
A ser feito nesse PR #64:

- Resolver os to do que estão comentados no código.

A ser feito em um novo PR de **refatoração**, após esse #64 for mergeado:

- Atualizar os códigos que fazem comparações de igualdade entre variáveis/atributos do tipo `Color`. Não é um breaking change mas é legal fazer essa atualização, já que não vai ser mais necessário comparar separadamente os 3 valores (`red`,  `green` e `blue`) entre duas variáiveis/atributos do tipo `Color`, agora podemos fazer uma comparação direta;
- Adicionar os operadores **equal (==)** e **not equal (!=)** no `Pixel` e atualizar os códigos que fazem comparações de igualdade entre variáveis/atributos do tipo `Pixel` (mesmo caso do `Color`).

mateuseap commented on include/entity/scene.h:
```diff
@@ -0,0 +1,27 @@
+class Scene {
+private:
+    std::vector<Light> lights;
+    Color environmentColor; // QUESTION: Should this be a Color or a Pixel?
```
O que tu achas @kinhosz? Vamos seguir com `Color` ou `Pixel` aqui?

kinhosz replied:
Color. Minha justificativa é a seguinte:

Todas as cores passam pela fórmula de phong, envolvendo multiplicacoes, somas e etc... Trabalhar com o não normalizado vai dar mais trabalho.

Color onde tiver cor.
Pixel quando for printar na tela. Logo, o único lugar no projeto que faz sentido ter pixel até agora é no frame, o resto deveria ser color mesmo.

mateuseap replied:
De acordo, vou manter como color então.

mateuseap commented on include/entity/scene.h:
```diff
@@ -0,0 +1,27 @@
+private:
+    std::vector<Light> lights;
+    Color environmentColor; // QUESTION: Should this be a Color or a Pixel?
+    // TO DO: Add objects
```
Ainda não declarei os `Objects` aqui pois acho que é necessário subir todos para `main` primeiro, faz sentido @kinhosz? Além disso, joga aqui uma sugestão de como esses objetos podem ser declarados por obséquio.

kinhosz replied:
Exato. A gente vai aprender algumas otimizações, mas não vai ser cobrado aqui nesse projeto. Eu pensei em termos um::

```cpp
vector<Box> objects;
```
meio que assim, pq isso já englobaria todos os objetos possíveis. Precisamos estudar se isso é sintaticamente aceito em C++, e se não, ver uma forma de fazer, saca? já que os objetos terão métodos semelhantes.

mateuseap replied:
No lugar desse `std::vector<Box> objects` eu defini um `std::map<int, Box> objects`, assim como havíamos combinado.

mateuseap commented on include/entity/scene.h:
```diff
@@ -0,0 +1,27 @@
+    void setLights(std::vector<Light> lights);
+    void setEnvironmentColor(Color environmentColor);
+
+    //TO DO: Add objects getters and setters
+    //TO DO: Add intersection method
```
Marcando aqui +2 to do para não esquecer

mateuseap replied:
@kinhosz se tiver alguma ideia em relação ao método de interseção (`intersect`), lança a braba para eu já ir pensando a respeito :eyes:

kinhosz replied:
vamos criar o `intersect` em outro PR. Vou abrir uma issue com os detalhes dela, pq ela vai ser bem mais complexa e acho que dentro desse escopo acaba ficando muita coisa.

kinhosz replied:
sobre os getters e os setters, pode ficar a vontade. Só queria abrir a discussão para uma coisa:

`vale a pena ter todos os atributos da classe disponíveis para ser setados ou pegados?`

Por exemplo, o plane, faz sentido a gente setar um vetor normal? o vetor normal não deveria ser constante durante toda a vida de um plane? 

Fica essas dúvidas

mateuseap replied:
@kinhosz vamos abrir uma issue para o `intersect`, posso tocar isso depois que esse PR #64 subir para `main`, com a estrutura base do `Scene`.

mateuseap replied:
Em relação a essa questão dos getters/setters, realmente, dependendo da classe alguns atributos não fazem sentidos serem setados ou até mesmo "solicitados". A gente pode levantar o que não faz sentido ter um getter/setter e depois, ajustar tudo em um único PR, o que achas?

kinhosz commented on src/entity/scene.cpp:
```diff
@@ -0,0 +1,21 @@
+#include <entity/scene.h>
+
+Scene::Scene() : lights(std::vector<Light>()), environmentColor(Color()) {}
+
+Scene::Scene(std::vector<Light> lights, Color environmentColor) : lights(lights), environmentColor(environmentColor) {}
+
+std::vector<Light> Scene::getLights() const {
+    return this->lights;
+}
+
+Color Scene::getEnvironmentColor() const {
+    return this->environmentColor;
+}
+
+void Scene::setLights(std::vector<Light> lights) {
+    this->lights = lights;
+}
+
+void Scene::setEnvironmentColor(Color environmentColor) {
+    this->environmentColor = environmentColor;
+}
```
acho que `Scene` é um bom candidato para ter seus construtores mais limpos. 

* lights: podemos criar métods addLigth, removeLight. Permitindo ser mais dinâmico. Seria legal representar cada Light a um id dentro do scene, usando map talvez?

* camera: podemos ter mais de uma camera na cena, então os mesmos questionamentos de cima persistem aqui

* objects: add / remove object.

mateuseap replied:
Feito, agora temos um `addLight`, `removeLight`, `addObject` e `removeObject`. Em relação a questão das câmeras (que tu levantou num comentário mais abaixo), acho que faz mais sentido deixarmos as câmeras as cenas separadas mesmo, que no caso foi a sugestão de número 3 que tu deu.

kinhosz commented:
Comentei algumas coisas para levantar a discussão.

kinhosz commented:
Voltando a discussão da câmera e da cena.

* Opção 1: Uma cena pode possuir várias câmeras
* Opção 2: Uma câmera pode possuir várias cenas
* Opção 3: Câmeras e cenas são independentes entre si

Eu gosto mais da 3° opção, e tu? Eu entendo uma cena sendo tipo uma chunck do minecraft

mateuseap committed:
feat(include/entity/scene.h): modify attributes and getters/setters

mateuseap committed:
feat(src/entity/scene.cpp): update source code to match with the new header file

mateuseap committed:
feat(tests/entity/scene.cpp): update validation tests

mateuseap committed:
feat(light): add empty constructor

mateuseap committed:
fix(tests/entity/scene.cpp): adjust validation tests


fix(tests/entity/scene.cpp): adjust validation tests


fix(tests/entity/scene.cpp): adjust syntax error

mateuseap commented:
> Voltando a discussão da câmera e da cena.
> 
> * Opção 1: Uma cena pode possuir várias câmeras
> * Opção 2: Uma câmera pode possuir várias cenas
> * Opção 3: Câmeras e cenas são independentes entre si
> 
> Eu gosto mais da 3° opção, e tu? Eu entendo uma cena sendo tipo uma chunck do minecraft

Respondi mais acima, mas também gosto mais da opção 3.

mateuseap commented:
> A ser feito nesse PR #64:
> 
> * Resolver os to do que estão comentados no código.
> 
> A ser feito em um novo PR de **refatoração**, após esse #64 for mergeado:
> 
> * Atualizar os códigos que fazem comparações de igualdade entre variáveis/atributos do tipo `Color`. Não é um breaking change mas é legal fazer essa atualização, já que não vai ser mais necessário comparar separadamente os 3 valores (`red`,  `green` e `blue`) entre duas variáiveis/atributos do tipo `Color`, agora podemos fazer uma comparação direta;
> * Adicionar os operadores **equal (==)** e **not equal (!=)** no `Pixel` e atualizar os códigos que fazem comparações de igualdade entre variáveis/atributos do tipo `Pixel` (mesmo caso do `Color`).

Agora que o `Light` também tem os operadores **equal (==)** e **not equal (!=)**, acho válido fazer o mesmo que citei que deve ser feito em relação ao `Color`.

kinhosz commented on src/entity/scene.cpp:
```diff
@@ -0,0 +1,64 @@
+
+Scene::Scene() {
+    this->lights = std::map<int, Light>();
+    this->lightsCurrentIndex = this->lights.size() - 1;
+    this->environmentColor = Color(0, 0, 0);
+    this->objects = std::map<int, Box>();
+    this->objectsCurrentIndex = this->objects.size() - 1;
```
this->lights.size()
```suggestion
    this->lightsCurrentIndex = this->lights.size();
    this->environmentColor = Color(0, 0, 0);
    this->objects = std::map<int, Box>();
    this->objectsCurrentIndex = this->objects.size();
```

mateuseap replied:
@kinhosz feito! Precisei alterar também a lógica utilizada no `addLIght` e no `addObject`, dá uma olhada se tá tudo ok.

kinhosz commented on src/entity/scene.cpp:
```diff
@@ -0,0 +1,64 @@
+}
+
+Scene::Scene(const Color& environmentColor) {
+    this->lights = std::map<int, Light>();
+    this->lightsCurrentIndex = this->lights.size() - 1;
+    this->environmentColor = environmentColor;
+    this->objects = std::map<int, Box>();
+    this->objectsCurrentIndex = this->objects.size() - 1;
```
o mesmo aqui

kinhosz commented on src/entity/scene.cpp:
```diff
@@ -0,0 +1,64 @@
+}
+
+Scene::Scene(const std::map<int, Light>& lights, const Color& environmentColor, const std::map<int, Box>& objects) {
+    this->lights = lights;
+    this->lightsCurrentIndex = this->lights.size() - 1;
+    this->environmentColor = environmentColor;
+    this->objects = objects;
+    this->objectsCurrentIndex = this->objects.size() - 1;
```
aqui tbm

kinhosz reviewed (changes requested):
só mudar esse -1 aí, o resto tá tudo safe

mateuseap committed:
feat(src/entity/scene.cpp): modify lights and objects current indexes logic

kinhosz reviewed (approved):
LGTM
